core/cap-0063.md

```diff
@@ -0,0 +1,496 @@
+```
+CAP: 0063
+Title: Soroban In-memory Read Resource
+Working Group:
+    Owner: Garand Tyson <@SirTyson>
+    Authors: Garand Tyson <@SirTyson>
+    Consulted: Dmytro Kozhevin <@dmkozh>, Nicolas Barry <@MonsieurNicolas>
+Status: Draft
+Created: 2024-12-09
+Discussion: https://github.com/stellar/stellar-protocol/discussions/1585
+Protocol version: 23
+```
+
+## Simple Summary
+
+This proposal introduces a new resource type for Soroban reads, distinguishing between in-memory and
+disk reads. This also proposes automatic restoration for archived entries via `InvokeHostFunctionOp`.
+
+## Working Group
+
+As specified in the Preamble.
+
+## Motivation
+
+By distinguishing disk and in-memory reads, this proposal allows for significant increase in Soroban
+read limits. This distinction also allows for safe automatic entry restoration, significantly improving UX.
+
+### Goals Alignment
+
+This change is aligned with the goal of lowering the cost and increasing the scale of the network.
+
+## Abstract
+
+[CAP-0062](cap-0062.md) introduces partial State Archival, where evicted and live Soroban state is stored in separate databases.
+This separation allows live Soroban state to be efficiently cached entirely in memory, removing disc reads entirely for all
+live Soroban state. However, classic state and evicted entries are still subject to disk reads.
+
+This CAP introduces a new resource type for reads, distinguishing between disk reads and in-memory reads. By making this
+distinction at the protocol level, read limits for Soroban data can greatly increase. This also opens the door for other
+optimizations, such as a complete module cache for all live contracts.
+
+Additionally, this CAP introduces automatic restoration via `InvokeHostFunctionOp`, where any archived key present in the
+footprint is automatically restored. Initially in protocol 20, the state archival design was not solidified enough to
+enable automatic restoration, so an explicit restore operation was required. While this was not technically required in
+protocol 20, it was important for contract interfaces to reason properly about restoration such that when full state
+archival was introduce, it would not break preexisting deployments. Given that the full State Archival design has been
+mostly finalized in [CAP-0057](cap-0057.md), it is now appropriate to introduce automatic restoration.
+
+## Specification
+
+### XDR changes
+
+```
+--- a/Stellar-contract-config-setting.x
++++ b/Stellar-contract-config-setting.x
+@@ -26,28 +26,28 @@ struct ConfigSettingContractComputeV0
+ // Ledger access settings for contracts.
+ struct ConfigSettingContractLedgerCostV0
+ {
+-    // Maximum number of ledger entry read operations per ledger
+-    uint32 ledgerMaxReadLedgerEntries;
+-    // Maximum number of bytes that can be read per ledger
+-    uint32 ledgerMaxReadBytes;
++    // Maximum number of live Soroban entry read operations per ledger
++    uint32 ledgerMaxReadLiveSorobanEntries;
```

Comment:
Maybe we should name this in a bit more explicit and high-level fashion, like `diskReadLedgerEntries`, `inMemoryLedgerEntries`. This will also allow us to change what's actually on disk and in-memory without changing the names and semantics of these fields.